src/sparse/KokkosSparse_BlockCrsMatrix.hpp

```diff
@@ -0,0 +1,957 @@
+  /// <li> \c view.local_block_value(K, i, j) returns a nonconst reference
+  ///      to the value in the ith local row and jth local col 
+  ///      of the k-th block of the block-row </li>
+  /// <li> \c view.block(K) returns an unmanaged 2D strided Kokkos::View
+  ///      of the values of the k-th block of the block-row </li>
+  /// </ul>
+  ///
+  /// Users should not rely on the return type of this method.  They
+  /// should instead assign to 'auto'.
+  ///
+  KOKKOS_INLINE_FUNCTION
+  SparseBlockRowViewConst<BlockCrsMatrix> block_row_Const (const ordinal_type i) const {
+
+    const size_type start = graph.row_map(i); // total num blocks prior to this block-row
+    const ordinal_type count = static_cast<ordinal_type> (graph.row_map(i+1) - start); // num blocks in this row
+
+    if (count == 0) {
+      return SparseBlockRowViewConst<BlockCrsMatrix> (nullptr, nullptr, 1, 0);
+    } else {
+      return SparseBlockRowViewConst<BlockCrsMatrix> (values, graph.entries, blockDim(), count, start);
+    }
+  }
+
+private:
+  ordinal_type numCols_;
+  ordinal_type blockDim_; // TODO Assuming square blocks for now - add blockRowDim, blockColDim
+};
+
+//----------------------------------------------------------------------------
+//----------------------------------------------------------------------------
+
+// Assumes rows is the row_map pointer for the BlockCrsMatrix graph (i.e. cum sum of number of blocks)
+//         cols is the entries pointer for the BlockCrsMatrix graph (colidx for block-row blocks)
+//         annz is the total number of non-zeros in the CrsMatrix (blockDim*blockDim*numBlocks)
+template< typename ScalarType , typename OrdinalType, class Device, class MemoryTraits, typename SizeType >
+void
+BlockCrsMatrix<ScalarType , OrdinalType, Device, MemoryTraits, SizeType >::
+ctor_impl (const std::string &label,
+           const OrdinalType nrows,
+           const OrdinalType ncols,
+           const size_type annz,
+           ScalarType* val,
+           OrdinalType* rows,
+           OrdinalType* cols,
+           const OrdinalType blockDim)
+{
+  std::string str = label;
+  values = values_type (str.append (".values"), annz);
+
+  numCols_ = ncols;
+  blockDim_ = blockDim;
+
+  // FIXME (09 Aug 2013) CrsArray only takes std::vector for now.
+  // We'll need to fix that.
+  std::vector<int> row_lengths (nrows, 0);
+
+  // FIXME (mfh 21 Jun 2013) This calls for a parallel_for kernel.
+  for (OrdinalType i = 0; i < nrows; ++i) {
+    row_lengths[i] = rows[i + 1] - rows[i];
+  }
+
+  str = label;
+  graph = Kokkos::create_staticcrsgraph<staticcrsgraph_type> (str.append (".graph"), row_lengths);
+  typename values_type::HostMirror h_values = Kokkos::create_mirror_view (values);
+  typename index_type::HostMirror h_entries = Kokkos::create_mirror_view (graph.entries);
+
+  // FIXME (mfh 21 Jun 2013) This needs to be a parallel copy.
+  // Furthermore, why are the arrays copied twice? -- once here, to a
```

Comment:
Let us avoid moving FIXME from one to another. Let us fix it or if it is low priority just make it a comment.